src/ch/ch_driver.c

```diff
@@ -0,0 +1,938 @@
+    virDomainPtr dom = NULL;
+
+    chDriverLock(driver);
+    vm = virDomainObjListFindByName(driver->domains, name);
+    chDriverUnlock(driver);
+
+    if (!vm) {
+        virReportError(VIR_ERR_NO_DOMAIN,
+                       _("no domain with matching name '%s'"), name);
+        goto cleanup;
+    }
+
+    if (virDomainLookupByNameEnsureACL(conn, vm->def) < 0)
+        goto cleanup;
+
+    dom = virGetDomain(conn, vm->def->name, vm->def->uuid, vm->def->id);
+
+ cleanup:
+    virDomainObjEndAPI(&vm);
+    return dom;
+}
+
+static virDomainPtr chDomainLookupByUUID(virConnectPtr conn,
+                                         const unsigned char *uuid)
+{
+    virCHDriverPtr driver = conn->privateData;
+    virDomainObjPtr vm;
+    virDomainPtr dom = NULL;
+
+    chDriverLock(driver);
+    vm = virDomainObjListFindByUUID(driver->domains, uuid);
+    chDriverUnlock(driver);
+
+    if (!vm) {
+        char uuidstr[VIR_UUID_STRING_BUFLEN];
+        virUUIDFormat(uuid, uuidstr);
+        virReportError(VIR_ERR_NO_DOMAIN,
+                       _("no domain with matching uuid '%s'"), uuidstr);
+        goto cleanup;
+    }
+
+    if (virDomainLookupByUUIDEnsureACL(conn, vm->def) < 0)
+        goto cleanup;
+
+    dom = virGetDomain(conn, vm->def->name, vm->def->uuid, vm->def->id);
+
+ cleanup:
+    virDomainObjEndAPI(&vm);
+    return dom;
+}
+
+static int
+chDomainGetState(virDomainPtr dom,
+                 int *state,
+                 int *reason,
+                 unsigned int flags)
+{
+    virDomainObjPtr vm;
+    int ret = -1;
+
+    virCheckFlags(0, -1);
+
+    if (!(vm = chDomObjFromDomain(dom)))
+        goto cleanup;
+
+    if (virDomainGetStateEnsureACL(dom->conn, vm->def) < 0)
+        goto cleanup;
+
+    *state = virDomainObjGetState(vm, reason);
+    ret = 0;
+
+ cleanup:
+    virDomainObjEndAPI(&vm);
+    return ret;
+}
+
+static char *chDomainGetXMLDesc(virDomainPtr dom,
+                                unsigned int flags)
+{
+    virCHDriverPtr driver = dom->conn->privateData;
+    virDomainObjPtr vm;
+    char *ret = NULL;
+
+    virCheckFlags(VIR_DOMAIN_XML_COMMON_FLAGS, NULL);
+
+    if (!(vm = chDomObjFromDomain(dom)))
+        goto cleanup;
+
+    // if (virDomainGetXMLDescEnsureACL(dom->conn, vm->def, flags) < 0)
+    //     goto cleanup;
+
+    ret = virDomainDefFormat(vm->def, driver->xmlopt,
+                             virDomainDefFormatConvertXMLFlags(flags));
+
+ cleanup:
+    virDomainObjEndAPI(&vm);
+    return ret;
+}
+
+static int chDomainGetInfo(virDomainPtr dom,
+                           virDomainInfoPtr info)
+{
+    virDomainObjPtr vm;
+    int ret = -1;
+
+    if (!(vm = chDomObjFromDomain(dom)))
+        goto cleanup;
+
+    if (virDomainGetInfoEnsureACL(dom->conn, vm->def) < 0)
+        goto cleanup;
+
+    info->state = virDomainObjGetState(vm, NULL);
+
+    info->cpuTime = 0;
+
+    info->maxMem = virDomainDefGetMemoryTotal(vm->def);
+    info->memory = vm->def->mem.cur_balloon;
```

Comment:
Why is the `cur_balloon` considered as the memory value?

Reply:
https://libvirt.org/formatdomain.html#memory-allocation is probably the best explanation, qemu and vmware seem to do about the same thing. Is this a feature that is supported in Cloud-Hypervisor (allocating less than the maximum memory initially and increasing up to the maximum)?

Regardless I think the code is correct as is.